fix regression in definition of beforeunload prompts caused by #1040 / #1518

jtnord opened the pull request:
The definition of what happens to a prompt created from a `beforeunload` event unintentionally changed from being implicitly dismissed from explicit navigation to implicit navigation.

the link used to go to "Go" but this was renamed in #1040 leaving a dead link.
This link was "fixed" in #1518 but instead of fixing the link to how it was (by linking to the new `"Navigate To"` which was renamed earlier) it linked to the default definition of `navigate` which is vastly different and covers all navigation implicit or explicit.

fixes #1294


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/jtnord/webdriver/pull/1573.html" title="Last updated on Feb 22, 2021, 2:35 PM UTC (5d22ffa)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/1573/90b179e...jtnord:5d22ffa.html" title="Last updated on Feb 22, 2021, 2:35 PM UTC (5d22ffa)">Diff</a>

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/w3c/webdriver/1573)
<!-- Reviewable:end -->

jtnord committed:
clarify the bahaviour of beforeunloaded prompts

The wording of the current entry implies to some that promots from
beforeunloaded are always dismissed and thus the promot is in effect
never shown and the user will have no chance to interact with it.

This interpretation would lead it to be impossible to interact with them
and test them - which seems to be the purpose of the webdriver spec, so
this would imply that that interpretation is incorrect.

This clarifies the wording to say that the promot is dismissed on
*subsequent* navigations, thus if a prompt was shown and a navigation
attempt was performed then it would be dismissed, however if the cause
of the alter was a navigation then the promot will be shown so that the
user can interact with it.

fixes #1294

jgraham reviewed (changes requested):
I don't think this is a clarification; it's a total change in the model. As written it implies that something like closing a window can fail, but subsequent calls will succeed. That would need much more support in the spec than this patch provides.

If there's a desire to change the way the protocol works with beforeUnload prompts, the first question is what existing implementations do. If current implementations always follow the current spec it seems likely that unilateral change in the spec will break tests. If there are various behaviours there might be more flexibility to change things without breaking users.

(FWIW I would expect this kind of thing to be fixed in the BiDi version of WebDriver, and clients will get to implement whatever behaviour they like).

jtnord commented:
> If there's a desire to change the way the protocol works with `beforeUnload` prompts, the first question is what existing implementations do. 

Chrome - allows the user to handle them
Previous versions of firefox (not sure if this is firefox or a geckodriver change) allowed the user to handle them
Firefox/geckodriver latest releases - does not allow the user to handle them.

> If current implementations always follow the current spec it seems likely that unilateral change in the spec will break tests.

See above - this is not the case.

> If there are various behaviours there might be more flexibility to change things without breaking users.

Potentially the issue at hand is that the link to `navigate` includes all forms of navigation (even ones performed by `click` on a `WebElement` rather than just explicit ones (`forward` / `back` `navigate to``) ?

jtnord commented:
> Potentially the issue at hand is that the link to navigate includes all forms of navigationby linking to dfn-navigation (even ones performed by click on a WebElement rather than just explicit ones (forward / back / navigate to) by linking to navigation (section 10)?

which could as I understand it have been introduced in 321fff4e12d80a745828f7c711fdc18932cca244 because prior to that change there was no definition of `navigation` and thus IIUC the link would have taken you to the section header?

jtnord commented:
incorrect link in the above.  It did indeed used to link not to `dfn-navigation` but `go`   (which was `explicit` navigation triggered by webdriver and not by implicit navigation caused by a `click` on an element.

33c1f38f5353ffd3033179bd698b2adf84f41e86

jtnord committed:
Fix the incorrect link to navigation

The "navigation" used in versions of the spec from
33c1f38f5353ffd3033179bd698b2adf84f41e86 used to link to "Go".

The previous commit attempted to fix this by removing the link and
linking to the generic navigation definition however this was not the
intent.

SO this was originally introduced in #1040

this fixes this and so now correctly shows when prompts from
beforeunload should be implicitly dismissed

jtnord commented on index.html:
```diff
@@ -9730,9 +9730,8 @@ <h2>User prompts</h2>
- are <a>dismissed</a> implicitly upon <a>navigation</a>
- or <a>close window</a>,
- regardless of the defined <a>user prompt handler</a>.
+ are <a>dismissed</a> implicitly upon a <a data-lt="Navigate To">navigation</a>
```
#1040 changed `go` to navigate to and broke the link.
#1518 then tried to fix warnings in invalid links but linked to the wrong thing.

jtnord commented:
@jgraham have you had a chance to look and (in)validate my latest analysis here?

Given that Mozilla Firefox / geckodriver is now implementing this "as written" and differently to other browsers based seems pretty scary to me.

jtnord commented:
I realise that #1518 changed `Go (navigate to)` to `navigate` in multiple places so this is a wider issue than just here.

jtnord commented on index.html:
```diff
@@ -9730,9 +9730,8 @@ <h2>User prompts</h2>
- are <a>dismissed</a> implicitly upon <a>navigation</a>
- or <a>close window</a>,
- regardless of the defined <a>user prompt handler</a>.
+ are <a>dismissed</a> implicitly upon a <a data-lt="Navigate To">navigation</a>
+ or a <a>close window</a>, regardless of the defined <a>user prompt handler</a>.
```
this makes things less clear.  `Navigate To` is a webdriver spec - (it is a POST) so can never be WebDriver Initiated - unless webdriver is calling itself?!  (in which case it would seem to be implementation specific what would or would not end up calling this (would back actually get some history then `Navigate To` in implenentation X and implementation Y just use some `back()` like functionaltity provided?
